### PR TITLE
vm: replace custom result-like types with `Result`

### PR DESCRIPTION
**Pull request by zerbina**
`VmGenResult` and `ExecutionResult` are both kept as aliases.

In addition:
- the handling of the result values is simplified/cleaned up a bit
- errors are always wrapped in `nkError` nodes and no longer reported
 during result unpacking

### Future work:
- `evalStaticStmt`  and `setupCompileTimeVar` need to further propagate
 the errors instead of `localReport`-ing them

### CODE REVIEW

File: compiler/vm/vm.nim [hunk 1]
```diff
@@ -3319,51 +3304,49 @@ proc execute(c: var TCtx, start: int, frame: sink TStackFrame; cb: proc(c: TCtx,
...
-  ## new `nkError` wrapping `node`. Otherwise, sets `start` to the instruction
-  ## offset given by the result and returns `nil`
-  if r.success:
-    result = nil
-    start = r.start
+template returnOnErr(res: VmGenResult, config: ConfigRef, node: PNode): int =
+  ## Unpacks the vmgen result. If the result represents an error, exits the
+  ## calling function by either returning a new `nkError` wrapping `node`
+  ## or, if the calling function has no return value, via
+  ## `handleReport`+`return`.
```
Comment (saem) [inline compiler/vm/vm.nim]: Can we not use an nkError in both cases except one is an empty?

File: compiler/vm/vm.nim [hunk 2]
```diff
@@ -3319,51 +3304,49 @@ proc execute(c: var TCtx, start: int, frame: sink TStackFrame; cb: proc(c: TCtx,
...
-  ## directly (via `handleReport`)
-  if res.success:
-    result = res.result
+  if res.isOk:
+    result = res.take
     if node != nil and result.info.line < 0:
       result.info = node.info
   else:
-    let errKind = res.report.kind
+    let err = res.takeErr
```
Comment (saem) [inline compiler/vm/vm.nim]: Nonblocking style remark: as you go can you favour let and var blocks?

**Review by saem — COMMENTED** [review]
This is super exciting!

We're on the cusp of being able to compile and instead of the current sem approach where we get a boolean, we can instead get the AST that worked out is an nkError!

File: compiler/vm/vm.nim (hunk 1 above)
Comment (zerbina) [reply compiler/vm/vm.nim]: `evalStmt` now also return a `PNode`, making the special `handleReport`+`return` logic here unnecessary.

**Comment by zerbina** [comment]
Fixed the failing tests (the VM stack-trace wasn't printed anymore) and also addressed the style remarks.

*Edit:* @saem If the tests pass (they should) and you're okay with the adjustments, feel free to bors already

**Comment by saem** [comment]
heck yeah, bors away! :D

**Pull request merged**